src/main/java/com/greghaskins/spectrum/model/Singleton.java

```diff
@@ -0,0 +1,26 @@
+package com.greghaskins.spectrum.model;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * The singleton pattern. An abstraction of a single object which can be created and
+ * destroyed. This is a twist on the usual singleton pattern, in that normally the
+ * singleton stays around indefinitely. This version is also thread safe.
+ */
+public class Singleton<T> implements Supplier<T> {
```

Comment:
My aim is to make it really thread safe in #61 for now, I'm just laying foundations.

I'm going to change the singleton stuff. You're right. It's got surprise in it. I'm going to claim that I am right to have chosen the name, since a Singleton is actually capable of abstracting the correct creation of an object and having more than one if you're doing things across threads.

In fact, though. The `Variable` needs to be threadsafe, and it seems like `let` and `Variable` ought to share more implementation. Watch this space.